File: python/queen2.py
from __future__ import print_function
import time
import pprint

def check(m, depth):
    if depth < 1:
        return True
    x = m[depth]
    for i in range(depth):
        dx = m[i] - x
        dy = depth - i
        if dx == 0:
            return False
        elif dx == dy or dx == -dy:
            return False
    return True

def search(size):
    m = [0] * size
    count = 0
    top = 0
    while 1:
        succeed = False
        if check(m, top):
            if top == size - 1:
                count += 1
            succeed = True
        if (not succeed) or (top == size - 1):
            while top >= 0:
                m[top] += 1
                if m[top] < size:
                    break
                top -= 1
            if top < 0:
                break
        else:
           top += 1
           m[top] = 0
    return count

def benchmark(times, func):
    import time
    t = time.time()
    for i in range(times):
        func()
    t = time.time() - t
    print('benchmark time: %.3f'%t)

print(search(8))

N = 10
benchmark(N, lambda: search(10))
benchmark(N, lambda: search(10))
benchmark(N, lambda: search(10))

File: lua/queen2.lua
function check(m, depth)
	if depth < 1 then
		return true
	end
	local x = m[depth]
	for i = 0, (depth - 1) do
		local dx = m[i] - x
		local dy = depth - i
		if dx == 0 then 
			return false
		elseif dx == dy or dx == -dy then
			return false
		end
	end
	return true
end

function search(size)
	local m = {}
	for i = 0, size do
		m[i] = 0
	end
	local count = 0
	local top = 0
	while 1 do
		local succeed = false
		if check(m, top) then
			if top == size - 1 then
				count = count + 1
			end
			succeed = true
		end
		if (not succeed) or (top == size - 1) then
			while top >= 0 do
				m[top] = m[top] + 1
				if m[top] < size then
					break
				end
				top = top - 1
			end
			if top < 0 then
				break
			end
		else
			top = top + 1
			m[top] = 0
		end
	end
	return count
end

function benchmark(times, func)
	local t = os.clock()
	for i = 1, times do
		func()
	end
	t = os.clock() - t
	print(string.format("benchmark time: %.3f", t))
end

print(search(8))

N = 10
benchmark(N, function () search(10) end)
benchmark(N, function () search(10) end)
benchmark(N, function () search(10) end)

